app/utils/database.py

```diff
@@ -148,6 +171,50 @@ class Herd(BaseModel):
     longitude = FloatField(null=True)
     coordinates_privacy = CharField(15, null=True)
 
+    def filtered_dict(self, user=None):
+        """
+        Returns the model data filtered by the access level of the given `user`.
+        """
+
+        access_level = 'public'
+        if user and user.is_admin:
+            access_level = 'private'
+        elif user:
+            for role in user.privileges:
+                if role['level'] in ['specialist', 'manager']:
+                    if role['genebank'] == self.genebank.id:
+                        access_level = 'private'
+                        break
+                elif role['level'] == 'owner':
+                    if role['herd'] == self.id:
+                        access_level = 'private'
+                        break
+                    access_level = 'authenticated'
+
+        data = self.as_dict()
+
+        # prune system data
+        del data['email_verified']
+
+        levels = ['public', 'authenticated', 'private']
+        # prune data according to access level
+        for field in [f for f in data.keys() if f.endswith('_privacy')]:
+            # remove values if access_level is less than required
+            field_level = data[field] or 'private'
+            if levels.index(access_level) < levels.index(field_level):
+                if field == 'coordinates_privacy':
+                    del data['latitude']
+                    del data['longitude']
+                else:
+                    target_field = field[:-8]
+                    del data[target_field]
+            # remove the access level value if the user doesn't have private
+            # access
+            if access_level != levels.index('private'):
```

Comment:
Already tested by the handling at L204 so this should go away?

Reply:
no. line 204 checks that you have permission to access the value field, which can vary in required permission, while this line removes the permission field which should only be accessible if you have specifically private permissions.
like, a field could have permission level "authenticated", and that could pass the check on line 204, but wouldn't pass 2013.

Reply:
I don't think I understand the logic then, but in addition: this compares the access_level (string) to the index in the list (integer).

But with a straight comparison if access_level is private, it would remove fields marked as public (that thus should be available to everybody)?

Reply:
This is fixed by using string comparison in #73